### PR TITLE
Synchronizing joinSession with TreesLatest to make opening sharing link that require redemption work

### PR DESCRIPTION
**Pull request by vladsud**
Issue https://github.com/microsoft/FluidFramework/issues/4954
Very hacky solution, but I'm not sure we can do better here.
Issue has more description on why and how.

Requires UTs to be added, but raising to get feedback if this is the right direction / logic.

Some background info:
All fetch calls in driver go through EpochTracker. This is used to detect version bump (epoch change) and properly fail in case client logic does not match server knowledge about file.

Note: Based on my analyzes, the chance of hitting it in Bohemia is zero. That's because we do not return container back until we got snapshot, and container always is started paused.

@sumedhb1995, can you please run it through your scenario and report if it addresses the problem?

### CODE REVIEW

**Comment by sumedhb1995** [comment]
Adding in UTs in this PR: https://github.com/vladsud/FluidFramework/pull/1/files

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 1]
```diff
@@ -229,5 +233,50 @@ export class EpochTracker {
...
+        try {
+            return await super.fetchAndParseAsJSON<T>(url, fetchOptions, fetchType, addInBody);
+        } catch (error) {
+            // Only handling here treesLatest. If createFile failed, we should never try to do joinSession.
+            // Similar, if getVersions failed, we should not do any further storage calls.
+            // So treesLatest is the only call that can have parallel joinSession request.
+            if (fetchType === "treesLatest") {
+                this.treesLatestDeferral.reject(error);
+            }
+            if (fetchType !== "joinSession" || error.statusCode !== 404 || !completed) {
```
Comment (sumedhb1995) [inline packages/drivers/odsp-driver/src/epochTracker.ts:271]:
```suggestion
            if (fetchType !== "joinSession" || error.statusCode !== 404 || completed) {
```
Switching this to completed to throw the error if we've already attempted a retry in the past.
This successfully initiates the retry in the 3p prototype scenario

**Review by sumedhb1995 — APPROVED** [review]

**Comment by jatgarg** [comment]
>         const epochTracker = new EpochTracker(new LocalPersistentCacheAdapter(this.persistedCache), logger2);

Here also use EpochTrackerWithRedemption.

---
Refers to: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:73 in 751d8b5. [](commit_id = 751d8b54d6c99275a8ade21c6e7dae6899246d18, deletion_comment = False)

File: packages/drivers/odsp-driver/src/epochTracker.ts [hunk 2]
```diff
@@ -240,5 +245,63 @@ export class EpochTracker {
...
+            if (fetchType === "treesLatest") {
+                this.treesLatestDeferral.reject(error);
+            }
+            if (fetchType !== "joinSession" || error.statusCode < 401 || error.statusCode > 404 || completed) {
+                throw error;
+            }
+        }
+
+        // It is joinSession failing with 401..404 error
+        // Repeat after waiting for treeLatest succeeding (of fail if it fails).
```
Comment (jatgarg) [inline packages/drivers/odsp-driver/src/epochTracker.ts:289]: "of" -> "or"

**Review by jatgarg — APPROVED** [review]
:shipit:

**Comment by vladsud** [comment]
> Here also use EpochTrackerWithRedemption.
I chose not to do it. Reason - while design of EpochTrackerWithRedemption ensures it should work for create file workflow, it's not needed here (there is no need to redeem sharing link and there is no sharing link to start with), and using it just adds more risk to the product (as waiting / timeout is not zero risk).

**Review by kj9lj — APPROVED** [review]

**Review by kj9lj — APPROVED** [review]

**Pull request merged**